src/xkbcomp/action.c

```diff
@@ -705,7 +705,7 @@ HandlePrivate(struct xkb_context *ctx, const struct xkb_mod_set *mods,
                 return false;
             }
 
-            strncpy((char *) act->data, str, sizeof(act->data));
```

Comment:
cc @fooishbar so he can confirm this

I dug into this a bit more and yes, the `strncpy()` seems to be correct. What's confusing is that we have a `data[7]` array and `strncpy` up to 7 bytes into it. This does result in a non-null terminated string when you set a action to 7 bytes: `Private(type=0x86, data="1234567")`.

All the current strings in xkeyboard-config at least are 6 bytes only but there's only [one suggestion](http://pascal.tsu.ru/en/xkb/gram-action.html) that strings should only be 6 bytes. I couldn't find anything else but nothing even referred to the 7 byte size.

The server handles it correctly either way, it allocates size + 1 and forcibly null-terminates it in `XkbDDXPrivate()`. Does anything else use this? Can't find anything in mutter or gnome-shell but not sure I'm grepping for the right thing. libxkbcommon itself seems to ignore that action type too.

It doesn't matter for the conversion from x11 keymap to string because strings are always decomposed into the numeric value-per-index anyway.

So the two options here are ignoring coverity's complaint (easy) or forcing the string length to a max of 6 (safer, but with the potential to break things). Any opinions?

Reply:
We probably shouldn't limit it to 6.

Maybe this will placate Coverity?

```diff
diff --git a/src/xkbcomp/action.c b/src/xkbcomp/action.c
index 5f1fd65..9f1f05b 100644
--- a/src/xkbcomp/action.c
+++ b/src/xkbcomp/action.c
@@ -698,14 +698,17 @@ HandlePrivate(struct xkb_context *ctx, const struct xkb_mod_set *mods,
 
             str = xkb_atom_text(ctx, val);
             len = strlen(str);
-            if (len < 1 || len > 7) {
+            if (len < 1 || len > sizeof(act->data)) {
                 log_warn(ctx,
-                         "A private action has 7 data bytes; "
-                         "Illegal data ignored\n");
+                         "A private action has %lu data bytes; "
+                         "Illegal data ignored\n",
+                         (unsigned long) sizeof(act->data));
                 return false;
             }
 
-            strncpy((char *) act->data, str, sizeof(act->data));
+            memset(act->data, 0, sizeof(act->data));
+            /* Not necessarily NUL-terminated. */
+            memcpy(act->data, str, len);
             return true;
         }
         else {
```

Reply:
I'd use `"%zd` for `sizeof()` but tbh, if we're leaving it like this anyway, we should just add a comment to indicate it's intentional and mark coverity as false positive. No need to add extra code. I'll update the commit to just add the comment line and re-push

Reply:
It's 6 bytes because that's how much space there is in the old wire protocol. We should enforce that limit so we don't cut X11 adrift.

Reply:
The wire protocol has size 7 too, no?

```
typedef struct _xkbActionWireDesc {
    CARD8       type;
    CARD8       data[7];
} xkbActionWireDesc;
#define sz_xkbActionWireDesc    8
```